internal/cms/signed_test.go

```diff
@@ -0,0 +1,418 @@
+
+package cms
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"reflect"
+	"testing"
+	"time"
+)
+
+func TestVerifySignedData(t *testing.T) {
+	ctx := context.Background()
+	// parse signed data
+	sigBytes, err := os.ReadFile("testdata/TimeStampToken.p7s")
+	if err != nil {
+		t.Fatal("failed to read test signature:", err)
+	}
+	signed, err := ParseSignedData(sigBytes)
+	if err != nil {
+		t.Fatal("ParseSignedData() error =", err)
+	}
+
+	// basic check on parsed signed data
+	if got := len(signed.Certificates); got != 4 {
+		t.Fatalf("len(Certificates) = %v, want %v", got, 4)
+	}
+	if got := len(signed.SignerInfos); got != 1 {
+		t.Fatalf("len(Signers) = %v, want %v", got, 1)
+	}
+
+	// verify with no root CAs and should fail
+	roots := x509.NewCertPool()
+	opts := x509.VerifyOptions{
+		Roots:       roots,
+		KeyUsages:   []x509.ExtKeyUsage{x509.ExtKeyUsageTimeStamping},
+		CurrentTime: time.Date(2022, 1, 1, 0, 0, 0, 0, time.UTC),
+	}
+	if _, err := signed.Verify(ctx, opts); err == nil {
+		t.Errorf("ParseSignedData.Verify() error = %v, wantErr %v", err, true)
+	} else if vErr, ok := err.(VerificationError); !ok {
+		t.Errorf("ParseSignedData.Verify() error = %v, want VerificationError", err)
+	} else if _, ok := vErr.Detail.(x509.UnknownAuthorityError); !ok {
+		t.Errorf("ParseSignedData.Verify() VerificationError.Detail = %v, want UnknownAuthorityError", err)
+	}
+
+	// verify with proper root CA
+	rootCABytes, err := os.ReadFile("testdata/GlobalSignRootCA.crt")
+	if err != nil {
+		t.Fatal("failed to read root CA certificate:", err)
+	}
+	if ok := roots.AppendCertsFromPEM(rootCABytes); !ok {
+		t.Fatal("failed to load root CA certificate")
+	}
+	verifiedCertChains, err := signed.Verify(ctx, opts)
+	if err != nil {
+		t.Fatal("ParseSignedData.Verify() error =", err)
+	}
+	if !reflect.DeepEqual(verifiedCertChains[0], signed.Certificates) {
+		t.Fatalf("ParseSignedData.Verify() = %v, want %v", verifiedCertChains, signed.Certificates[:1])
+	}
+}
+
+func TestParseSignedData(t *testing.T) {
+	t.Run("invalid berData", func(t *testing.T) {
+		_, err := ParseSignedData([]byte("invalid"))
+		if err == nil {
+			t.Fatal("ParseSignedData() error = nil, wantErr true")
+		}
+	})
+
+	t.Run("invalid contentInfo", func(t *testing.T) {
+		_, err := ParseSignedData([]byte{0x30, 0x00})
+		if err == nil {
+			t.Fatal("ParseSignedData() error = nil, wantErr true")
+		}
+	})
+
+	t.Run("content type is not signed data", func(t *testing.T) {
+		_, err := ParseSignedData([]byte{
+			0x30, 0x12, 0x06, 0x0b, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x09, 0x10, 0x01, 0x06, 0xa0, 0x03, 0x04, 0x01, 0x78,
+		})
+		if err != ErrNotSignedData {
+			t.Errorf("ParseSignedData() error = %v, wantErr %v", err, ErrNotSignedData)
+		}
+	})
+
+	t.Run("invalid signed data content", func(t *testing.T) {
+		_, err := ParseSignedData([]byte{
+			0x30, 0x10, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x07, 0x02, 0xa0, 0x03, 0x04, 0x01, 0x78,
+		})
+		fmt.Println(err)
```

Comment:
I ran the unit test myself and found this super nit:
```suggestion
		
```

Reply:
Removed.